Enumerate all and only downloaded subdirectories

8kat opened the pull request:
This PR addresses item 2 of #58 by enumerating over every item in the source directory, allowing for the download contents to change significantly. There are a couple of design considerations here:

1. Should we only copy over subdirectories, or every entry in a directory? currently this PR does the latter. 
2. Should there be user-facing warnings if an expected subdirectory (e.g. plugins) is missing? What if there's a directory/file we normally wouldn't expect?

I also removed unitas_patchers_dir() since it was only used in source_dest_dirs.

8kat committed:
Remove duplicated directory in download path

8kat committed:
Merge pull request #1 from Eddio0141/main

jdlksaf

8kat committed:
formatting + remove comment

8kat committed:
expose copy_dir_all errors to end user

8kat committed:
add context again

8kat committed:
whitespace

8kat committed:
add check to see if source directory exists

8kat committed:
enumerate over unitas subdirectories

8kat committed:
remove dead code

8kat committed:
formatting

Eddio0141 commented:
Yeah now I see how inconsistent my UniTAS project release directory structure is.

I'm thinking of changing this so the directory would reflect what would end up on the game's directory, so the zip would contain the `BepInEx` directory which could be directly copied, which I can apply to all releases of UniTAS.

On future UniTAS releases the content of what ends up in the game directory could change, but it shouldn't extend beyond the game directory.

Example of this change [UniTAS_BepInEx_v0.5.0.zip](https://github.com/Eddio0141/unitas_setup_tool/files/12345580/UniTAS_BepInEx_v0.5.0.zip)

As for the second problem, I don't have plans right now to have things checked, but there could be a hash check in some way to make sure things are downloaded properly and if the folder has been changed and handle that accordingly. Probably would be done in another pr if there's some idea to handling this.

Eddio0141 reviewed (changes requested):
I've changed the UniTAS release to match the game's base directory.
Copying all content from inside the folder directly onto the game directory should work.

8kat committed:
reduce dest_dir to game_dir

8kat commented:
I just removed the following: 

```rust
let dest_dir = game_dir.join("BepInEx");
```
and replaced subsequent references to dest_dir with game_dir. 

Do we care about the fact with old versions of UniTAS the file structure wouldn't necessarily work? If it is then I think adding a versioning check to conditionally join BepInEx would be reasonable instead.

Eddio0141 commented:
No worries about that, I've updated all of the past release zips to match the current file structure, as they only contain either plugins or patchers.
As for the main branch and future releases, it will be following the same structure as well.

I tested the removal of the `game_dir.join("BepInEx")` and it works fine it seems so you can push that change.

8kat commented:
branch should have the changes in place now.

Eddio0141 reviewed (approved):
Looks good, thanks!
